### PR TITLE
Feature/gem ada features

### PR DESCRIPTION
**Pull request by guymaness2**
##### What does the change add or fix?

This update is focused on significantly enhancing the metric import capabilities of Bulldozer.

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

* Update Metric importing process to all new V2 style, dramatically improving both performance and versatility of metric importing.
* Added support for adding Note to occurrences with Attendance import.
* Minor tweaks to Schedule import logic to allow for tighter connection between imported schedules and metric partitions.

---------

##### Who reported, requested, or paid for the change?

Ada Bible Church

---------

##### Does this update or add options to the block UI?

no

---------

##### What files does it affect?

* Bulldozer.CSV/Bulldozer.CSV.csproj
* Bulldozer.CSV/CSVComponent.cs
* Bulldozer.CSV/HelperMethods.cs
* Bulldozer.CSV/Maps/Attendance.cs
* Bulldozer.CSV/Maps/Metrics.cs
* Bulldozer.CSV/Maps/Schedule.cs
* Bulldozer.CSV/Model/AttendanceCsv.cs
* Bulldozer.CSV/Model/MetricCsv.cs
* Bulldozer.CSV/Model/MetricValueCsv.cs
* Bulldozer.CSV/Model/ScheduleCsv.cs
* Bulldozer/Bulldozer.csproj
* Bulldozer/Model/AttendanceImport.cs
* Bulldozer/Model/MetricValuePartitionImport.cs

---------

##### Is it a breaking change for other versions/clients?

no

### CODE REVIEW

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 1]
```diff
@@ -334,5 +335,742 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+            var metricSourceTypes = DefinedTypeCache.Get( new Guid( Rock.SystemGuid.DefinedType.METRIC_SOURCE_TYPE ) ).DefinedValues;
+            var metricManualSource = metricSourceTypes.FirstOrDefault( m => m.Guid == new Guid( Rock.SystemGuid.DefinedValue.METRIC_SOURCE_VALUE_TYPE_MANUAL ) );
+            var entityTypes = EntityTypeCache.All().Where( e => e.IsEntity && e.IsSecured ).ToList();
+
+            var metricCategoryEntityTypeId = EntityTypeCache.Get<MetricCategory>( false, rockContext ).Id;
+
+            var metricCategories = categoryService.Queryable().AsNoTracking()
+                .Where( c => c.EntityTypeId == metricCategoryEntityTypeId ).ToList();
+            var existingCategoryForeignKeys = metricCategories.Select( c => c.ForeignKey ).ToList();
+            var defaultForeignKey = $"{this.ImportInstanceFKPrefix}^{metricCategoryEntityTypeId}_Metrics";
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Metrics.cs:358]:
```suggestion
            var defaultCategoryForeignKey = $"{this.ImportInstanceFKPrefix}^{metricCategoryEntityTypeId}_Metrics";
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 2]
```diff
@@ -334,5 +335,742 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+            // If only one match is returned, select it
+            // OR if multiple are returned but one matches an existing foreign key, select the first one to avoid duplicate foreign keys
+            if ( defaultMatchingCategories.Count == 1 || defaultMatchingCategories.Any( mc => existingCategoryForeignKeys.Any( k => mc.ForeignKey == k ) ) )
+            {
+                defaultMetricCategory = defaultMatchingCategories.First();
+            }
+
+            if ( defaultMetricCategory == null )
+            {
+                defaultMetricCategory = new Category
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:371]: Can this be consolidated into a reusable method?

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 3]
```diff
@@ -334,5 +335,742 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+                if ( matchingCategories.Count == 1 || matchingCategories.Any( mc => existingCategoryForeignKeys.Any( k => mc.ForeignKey == k ) ) )
+                {
+                    newmetricCategory = matchingCategories.First();
+                }
+
+                parentCategory = GetCategoriesByFKOrName( rockContext, parentForeignKey, category.ParentCategory, metricCategories ).FirstOrDefault();
+
+                if ( newmetricCategory == null )
+                {
+                    newmetricCategory = new Category
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:460]: Can this be consolidated into a reusable method?

**Comment by nateh777** [comment]
@guymaness2 need to make a metric category model due to issue we found with category/parent link.

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 4]
```diff
@@ -334,5 +335,742 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+                                .ToDictionary( k => k.Key, v => v.FirstOrDefault() );
+
+            var metricCsvsToProcess = this.MetricCsvList.Where( m => !metricLookup.ContainsKey( $"{this.ImportInstanceFKPrefix}^{m.Id}" ) );
+            var existingMetricCount = this.MetricCsvList.Count() - metricCsvsToProcess.Count();
+            if ( existingMetricCount > 0 )
+            {
+                this.ReportProgress( 0, $"{existingMetricCount} out of {this.MetricValueCsvList.Count()} Metric(s) from import already exist and will be skipped." );
+            }
+
+            foreach ( var metricCsv in this.MetricCsvList )
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:506]:
```suggestion
            foreach ( var metricCsv in metricCsvsToProcess )
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 5]
```diff
@@ -334,5 +335,742 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+            {
+                var foreignKey = $"{this.ImportInstanceFKPrefix}^{metricCsv.Id}";
+                if ( metricLookup.ContainsKey( foreignKey ) )
+                {
+                    continue;
+                }
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:512]:
```suggestion
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 6]
```diff
@@ -334,5 +335,742 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+
+                newMetrics.Add( newMetric );
+                metricLookup.Add( foreignKey, newMetric );
+            }
+
+            // Slice data into chunks and process
+            var workingMetricImportList = newMetrics.ToList();
+            var metricsRemainingToProcess = workingMetricImportList.Count;
+            var completedMetrics = 0;
+            var insertedGroups = new List<Group>();
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:550]:
```suggestion
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 7]
```diff
@@ -334,5 +335,742 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+
+            if ( missingMetricIds.Count > 0 )
+            {
+                LogException( $"MetricImport", $"The following metric ids were not found while trying to add imported metrics to categories:\r\n{string.Join( ", ", missingMetricIds )}" );
+            }
+
+            ReportProgress( 0, "Begin processing metric partitions records" );
+
+            var metricsWithPartitions = this.MetricCsvList.Where( m => m.Partition1Id.IsNotNullOrWhiteSpace() || m.Partition2Id.IsNotNullOrWhiteSpace() || m.Partition3Id.IsNotNullOrWhiteSpace() ).ToList();
+            metricLookup = new MetricService( rockContext )
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:627]: Do we need to reload this?

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 8]
```diff
@@ -334,5 +335,742 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+            var importedDateTime = RockDateTime.Now;
+            var metricPartitionsToInsert = new List<MetricPartition>();
+
+            foreach ( var metricCsv in metricCsvs )
+            {
+                var metric = metricLookup.GetValueOrNull( $"{this.ImportInstanceFKPrefix}^{metricCsv.Id}" );
+                var entityTypeName = string.Empty;
+                if ( metricCsv.Partition1Id.IsNotNullOrWhiteSpace() && !metricPartitionLookup.ContainsKey( $"{this.ImportInstanceFKPrefix}^{metricCsv.Id}_1" ) )
+                {
+                    entityTypeName = metricCsv.Partition1EntityTypeName;
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:687]: Let's try to consolidate this to use a common method somehow.

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 9]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+            if ( existingMetricValueCount > 0 )
+            {
+                this.ReportProgress( 0, $"{existingMetricValueCount} out of {this.MetricValueCsvList.Count()} Metric Value(s) from import already exist and will be skipped." );
+            }
+
+            var valuesWithInvalidMetricIds = newMetricValueCsvs.Where( mv => !metricLookup.ContainsKey( $"{this.ImportInstanceFKPrefix}^{mv.MetricId}" ) );
+            var metricValueCsvsToProcess = newMetricValueCsvs.Where( mv => metricLookup.ContainsKey( $"{this.ImportInstanceFKPrefix}^{mv.MetricId}" ) ).ToList();
+            if ( valuesWithInvalidMetricIds.Any() )
+            {
+                this.ReportProgress( 0, $"{metricValueCsvsToProcess.Count() - valuesWithInvalidMetricIds.Count()} Metric Value(s) from import have invalid MetricId values and will be skipped.." );
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:671]:
```suggestion
                this.ReportProgress( 0, $"{valuesWithInvalidMetricIds.Count()} Metric Value(s) from import have invalid MetricId values and will be skipped.." );
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 10]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+            
+            foreach( var metricPartitonCsv in metricValueCsvsWithPartition )
+            {
+                var metric = metricLookup.GetValueOrNull( $"{this.ImportInstanceFKPrefix}^{metricPartitonCsv.MetricId}" );
+                var metricValue = metricValueLookup.GetValueOrNull( $"{this.ImportInstanceFKPrefix}^{metricPartitonCsv.Id}" );
+                if ( metric == null || metricValue == null )
+                {
+                    continue;
+                }
+                if ( metricPartitonCsv.Partition1EntityId.IsNotNullOrWhiteSpace() )
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:760]: Refactorable to common method?

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 11]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+                }
+            }
+
+            //  Now We will process them by entitytype so we only have to use reflection once per entity type.
+
+            var campusEntityType = EntityTypeCache.Get( new Guid( Rock.SystemGuid.EntityType.CAMPUS ) );
+            var metricPartitionEntityTypes = metricPartitionLookup.Select( mp => mp.Value.EntityType ).DistinctBy( e => e.Id ).ToList();
+            foreach ( var entityType in metricPartitionEntityTypes )
+            {
+                var metricValuePartionsToInsert = new List<MetricValuePartition>();
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Metrics.cs:825]:
```suggestion
                var metricValuePartitonsToInsert = new List<MetricValuePartition>();
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 12]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+                            var newMetricValuePartition = new MetricValuePartition
+                            {
+                                MetricValueId = metricValuePartitionImport.MetricValue.Id,
+                                MetricPartitionId = metricValuePartitionImport.MetricPartition.Id,
+                                EntityId = campus.Id,
+                                ForeignKey = metricValuePartitionImport.ForeignKey
+                            };
+
+                            metricValuePartionsToInsert.Add( newMetricValuePartition );
+                        }
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:866]:
```suggestion
                          var campus = this.CampusesDict.GetValueOrNull( partitionCampusId.Value );
                          if ( campus == null )
                          {
                              errors += $"{DateTime.Now}, MetricValuePartition, Campus {partitionCampusId.Value} not found. MetricPartitionValue for {metricValuePartitionImport.CsvMetricValueId} metric value was skipped.\r\n";
                              continue;
                          }
                          var newMetricValuePartition = new MetricValuePartition
                          {
                              MetricValueId = metricValuePartitionImport.MetricValue.Id,
                              MetricPartitionId = metricValuePartitionImport.MetricPartition.Id,
                              EntityId = campus.Id,
                              ForeignKey = metricValuePartitionImport.ForeignKey
                          };

                          metricValuePartionsToInsert.Add( newMetricValuePartition );
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 13]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+                            }
+                            var newMetricValuePartition = new MetricValuePartition
+                            {
+                                MetricValueId = metricValuePartitionImport.MetricValue.Id,
+                                MetricPartitionId = metricValuePartitionImport.MetricPartition.Id,
+                                EntityId = campus.Id,
+                                ForeignKey = metricValuePartitionImport.ForeignKey
+                            };
+
+                            metricValuePartionsToInsert.Add( newMetricValuePartition );
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Metrics.cs:865]:
```suggestion
                            metricValuePartitionsToInsert.Add( newMetricValuePartition );
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 14]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+                            }
+                            var newMetricValuePartition = new MetricValuePartition
+                            {
+                                MetricValueId = metricValuePartitionImport.MetricValue.Id,
+                                MetricPartitionId = metricValuePartitionImport.MetricPartition.Id,
+                                EntityId = entity.Id,
+                                ForeignKey = metricValuePartitionImport.ForeignKey
+                            };
+
+                            metricValuePartionsToInsert.Add( newMetricValuePartition );
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Metrics.cs:907]:
```suggestion
                            metricValuePartitionsToInsert .Add( newMetricValuePartition );
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 15]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+                    }
+                }
+
+                if ( metricValuePartionsToInsert.Count > 0 )
+                {
+                    ReportProgress( 0, $"Begin processing { entityType.FriendlyName } type Metric Value Partitions." );
+                }
+
+                // Slice Metric Value Partition data into chunks and process
+                var workingMetricValuePartitionImportList = metricValuePartionsToInsert;
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Metrics.cs:919]:
```suggestion
                var workingMetricValuePartitionImportList = metricValuePartitionsToInsert;
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 16]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+
+                    if ( contextService != null )
+                    {
+                        MethodInfo qryMethod = contextService.GetType().GetMethod( "Queryable", new Type[] { } );
+                        var entityQry = qryMethod.Invoke( contextService, new object[] { } ) as IQueryable<IEntity>;
+                        var entityIdDict = entityQry
+                                            .Where( e => e.ForeignKey != null && e.ForeignKey.StartsWith( this.ImportInstanceFKPrefix + "^" ) )
+                                            .GroupBy( e => e.ForeignKey )
+                                            .ToDictionary( k => k.Key, v => v.FirstOrDefault() );
+
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Metrics.cs:889]:
```suggestion
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 17]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:890]:
```suggestion
                                            .ToDictionary( k => k.Key, v => v.FirstOrDefault() );

```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 18]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+                // Slice Metric Value Partition data into chunks and process
+                var workingMetricValuePartitionImportList = metricValuePartionsToInsert;
+                var metricValuePartitionsRemainingToProcess = workingMetricValuePartitionImportList.Count();
+                var completedMetricValuePartitions = 0;
+
+                while ( metricValuePartitionsRemainingToProcess > 0 )
+                {
+                    if ( completedMetricValuePartitions > 0 && completedMetricValuePartitions % ( this.DefaultChunkSize * 10 ) < 1 )
+                    {
+                        ReportProgress( 0, $"{completedMetricValuePartitions} Metric Values processed." );
```
Comment (guymaness2) [inline Bulldozer.CSV/Maps/Metrics.cs:927]:
```suggestion
                        ReportProgress( 0, $"{completedMetricValuePartitions} Metric Value Partitions processed." );
```

File: Bulldozer.CSV/Maps/Metrics.cs [hunk 19]
```diff
@@ -334,5 +335,615 @@ private void SaveMetrics( List<MetricValue> metricValues )
...
+                    }
+                }
+            }
+            if ( errors.IsNotNullOrWhiteSpace() )
+            {
+                LogException( null, errors, hasMultipleErrors: true );
+            }
+
+            return completedMetricValues;
+        }
```
Comment (nateh777) [inline Bulldozer.CSV/Maps/Metrics.cs:947]: Ensure this is at the right level of this method.

File: Bulldozer.CSV/Maps/Metrics.cs (hunk 13 above)
Comment (guymaness2) [reply Bulldozer.CSV/Maps/Metrics.cs:865]: Pulled suggestion into my larger, overlapping suggestion

File: Bulldozer.CSV/Maps/Metrics.cs (hunk 19 above)
Comment (guymaness2) [reply Bulldozer.CSV/Maps/Metrics.cs:914]: confirmed

**Comment by guymaness2** [comment]
@nateh777 I think this one is all wrapped up now, including new sample csv files that match the changes. They have been tested and verified with a full bulldozer run.

**Review by nateh777 — APPROVED** [review]
Needs to be tested on a 15.3 Database make sure there are no exceptions. And then we will cherry-pick to hotfix-16 and we'll test again on a 16.3 database and install.

**Comment by guymaness2** [comment]
@nateh777 successful test run against Rock v 15.3 database with sample csv files. Also running full migration test with a full db. It will take a couple hours to complete, but no issues so far. I feel comfortable saying this is good to merge.

**Pull request merged**